### ISSUE TITLE
[HW2] async 한 업데이트 Test 문제

### ISSUE THREAD

**Issue by DHdroid** [swsnu/swppfall2019#113]
ComponentDidMount에서 
        this.props.onGetArticle(this.props.match.params.id).then(res=>this.setState({
            ...this.state,
            title:this.props.storedArticle.title,
            content:this.props.storedArticle.content
        }));
처럼 then 함수를 이용한 State 업데이트를 했는데요, 이것이 mount 당시 바로 적용되지 않아 어려움을 겪고 있습니다.
지금은 
        for(var i=0; i<20; i++) {
            await component.update();
        }
로 업데이트를 반복적으로 수정하는 것으로 에러를 틀어막고 있는데, 좀 더 깔끔한 방법이 있을 것 같아 질문드립니다 ㅠㅠ..

**Comment by ktaebum** [comment]
onGetArticle 내부에서 `storedArticle`의 title과 content를 업데이트 해주면 되지 않을까요??

**Comment by DHdroid** [comment]
> onGetArticle 내부에서 `storedArticle`의 title과 content를 업데이트 해주면 되지 않을까요??

이게 storedArticle 은 제대로 된 값이 들어가는데 component state의 title과 content 가 업데이트가 안됩니다..ㅠㅠ

**Comment by ktaebum** [comment]
음 굳이 component state에 별도로 title과 content를 둬야할지는 잘 모르겠습니다.

render를 할 때 바로 가져와서 띄워도 된다고 생각합니다

**Comment by DHdroid** [comment]
editarticle 에 들어갈 때, storedArticle의 title과 content를  text area의 기본값으로 주기 위해서 state를 이용했습니다. storedArticle.title을 직접 기본값으로 주면 제대로 mount가 되지 않은 상태에서 render를 해서 기본값이 안들어가서, state의 title에 그 값을 넣고, state의 title과 content를 null 로 두고 non null이 되었을 때 input component를 활성화 시켜주는 방식을 택했습니다.
혹시 이런 경우에는 구현 자체를 뒤바꾸는 것 외에 update를 promise.then 이 끝난 시점으로 할 수 있는 방법은 없는 건가요??

**Comment by ktaebum** [comment]
아 edit article에 해당하는 내용이군요

저 부분은 `onGetArticle`을 하면 state transition이 되어 있어서
`then`이 아니라
바로 sequential하게 `setState`를 불러도 문제없이 수행 되지 않나요?

**Comment by DHdroid** [comment]
네 문제없이 수행되지만 기존의 증상과 똑같이 test 에서는 componentDidMount가 채 완료되기 전에 render가 돼버리는 것 같습니다. 혹시 몰라 componentDidMount 를 async 화 하고 전부 await를 걸어봤지만 결과는 같네요 ㅠㅠ

**Comment by ktaebum** [comment]
음 이게 test에 한해서라면 어짜피 그 부분은 신경쓰지 않고 (test에서 실제 axios call을 하진 않을테니)
임의로 storedArticle에 해당하는 state로 mockStore를 만들고 적절히 actionCreator를 spy한 다음에 테스트만 해도 되는 문제 아닌가요??

**Comment by DHdroid** [comment]
테스트 시에 component state가 요구되게 코드를 짜놔서요.. 일단 생각해보고 안되면 setTimeout 등으로 잠시 기다린 후 테스트를 진행하게 하도록 하겠습니다. 감사합니다!

**Comment by HaeTheong** [comment]
저도 비슷한 문제를 겪었는데 저는

it('test', done => {
	const com = (
        <Provider store={mockStore}>
          <ConnectedRouter history={history}>
            <클래스명 />
          </ConnectedRouter>
        </Provider>
      )
    const p = new Promise(resolve => {
      const component = mount(com);
      const instance = component.find(클래스명.WrappedComponent).instance();
      instance.componentDidMount();
      resolve();
    });
    p.then(() => {
      expect(~~).~~;
      done();
    });
  });

이렇게 instance.componentDidMount(); 를 비동기 처리해서 문제를 해결했습니다.

**Comment by DHdroid** [comment]
> 저도 비슷한 문제를 겪었는데 저는
> 
> ```
> it('test', done => {
>     const p = new Promise(resolve => {
> 		const com = (
> 			<Provider store={mockStore}>
>           		<ConnectedRouter history={history}>
>             		<클래스명 />
>           		</ConnectedRouter>
>         	</Provider>
>       	)
>       	const component = mount(com);
>       	const instance = component.find(클래스명.WrappedComponent).instance();
>       	instance.componentDidMount();
>       	resolve();
>     });
>     p.then(() => {
>       expect(~~).~~;
>       done();
>     });
> });
> ```
> 
> 이렇게 instance.componentDidMount(); 를 비동기 처리해서 문제를 해결했습니다.

앗 감사합니다!!
